### PR TITLE
minimal hook into loading necessary for Pkg3

### PR DESCRIPTION
**Pull request by StefanKarpinski**
This is a less disruptive alternative to https://github.com/JuliaLang/julia/pull/19910 and https://github.com/JuliaLang/julia/pull/19866. Doesn't really change anything, just makes how code gets loaded extensible via putting things that aren't Strings in `LOAD_PATH` and defining appropriate methods for `Base.load_hook`.

It does also remove searching for files relative to the current directory behavior, but I don't think that was used by anything except for the reflection code, I've now factored out.

### CODE REVIEW

File: base/initdefs.jl [hunk 1]
```diff
@@ -32,9 +32,9 @@ isinteractive() = (is_interactive::Bool)
 
 An array of paths (as strings) where the `require` function looks for code.
```
Comment (tkelman) [inline base/initdefs.jl]: needs some doc updating

File: test/loading.jl [hunk 2]
```diff
@@ -40,7 +40,23 @@ mktempdir() do dir
     end
 end
 
-let paddedname = "Ztest_sourcepath.jl"
-    filename = SubString(paddedname, 2, length(paddedname))
-    @test Base.find_in_path(filename) == abspath(paddedname[2:end])
+empty!(LOAD_PATH)
```
Comment (tkelman) [inline test/loading.jl]: should probably restore prior value when done?

File: base/reflection.jl [hunk 3]
```diff
@@ -720,6 +720,12 @@ Get the name of a generic `Function` as a symbol, or `:anonymous`.
 """
 function_name(f::Function) = typeof(f).name.mt.name
 
+function find_source_file(file::String)
```
Comment (tkelman) [inline base/reflection.jl]: does this get called on package code?

File: base/loading.jl [hunk 4]
```diff
@@ -80,54 +80,33 @@ else
...
     isfile_casesensitive(path) && return abspath(path)
     return nothing
 end
+load_hook(prefix::String, name::String, found::String) = found
+load_hook(prefix, name::String, found) =
+    throw(ArgumentError("unrecognized custom loader in LOAD_PATH: $prefix"))
 
-# `wd` is a working directory to search. defaults to current working directory.
-# if `wd === nothing`, no extra path is searched.
-function find_in_path(name::String, wd)
```
Comment (tkelman) [inline base/loading.jl]: Blink, PkgDev, and node-julia are using the 2 arg version of this. And ClobberingReload is using find_in_node_path. deprecate?

Comment (StefanKarpinski) [inline base/loading.jl]: Ok, will do.

File: base/reflection.jl (hunk 3 above)
Comment (tkelman) [inline base/reflection.jl]: calling edit on a method from a package?

**Review by tkelman — CHANGES_REQUESTED** [review]
The change to `find_source_file` breaks `edit("Example.jl")`. 2-arg `find_in_path`, and `find_in_node_path` should be deprecated.

**Comment by StefanKarpinski** [comment]
Closed in favor of https://github.com/JuliaLang/julia/pull/20120

**Pull request closed without merging**